From ffda7cfaf1f7acdc38bed15a877c62b47f9e515c Mon Sep 17 00:00:00 2001
From: Eliad Peller <eliad@wizery.com>
Date: Mon, 5 Mar 2012 17:09:55 +0200
Subject: [PATCH 57/58] BSS: Fix use-after-realloc

After reallocation of the bss struct, current_bss wasn't updated and
could hold an invalid pointer (which might get dereferenced later).

Update current_bss if the pointer was changed.

Signed-hostap: Eliad Peller <eliad@wizery.com>
intended-for: hostap-1
(cherry picked from commit eb37e085a4c17a7ebdf258d480c5f2c8a2ac7f08)
---
 wpa_supplicant/bss.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/wpa_supplicant/bss.c
+++ b/wpa_supplicant/bss.c
@@ -332,6 +332,8 @@ static void wpa_bss_update(struct wpa_su
 		nbss = os_realloc(bss, sizeof(*bss) + res->ie_len +
 				  res->beacon_ie_len);
 		if (nbss) {
+			if (wpa_s->current_bss == bss)
+				wpa_s->current_bss = nbss;
 			bss = nbss;
 			os_memcpy(bss + 1, res + 1,
 				  res->ie_len + res->beacon_ie_len);
